presto-memory-context/src/main/java/com/facebook/presto/memory/context/SimpleLocalMemoryContext.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory.context;
+
+import com.google.common.util.concurrent.ListenableFuture;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+import static com.facebook.presto.memory.context.MemoryReservationHandler.NOT_BLOCKED;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class SimpleLocalMemoryContext
+        implements LocalMemoryContext
+{
+    private final AggregatedMemoryContext parentMemoryContext;
+    @GuardedBy("this")
+    private long usedBytes;
+    @GuardedBy("this")
+    private boolean closed;
+
+    public SimpleLocalMemoryContext(AggregatedMemoryContext parentMemoryContext)
+    {
+        this.parentMemoryContext = requireNonNull(parentMemoryContext, "parentMemoryContext is null");
+    }
+
+    @Override
+    public synchronized long getBytes()
+    {
+        return usedBytes;
+    }
+
+    @Override
+    public synchronized ListenableFuture<?> setBytes(long bytes)
+    {
+        checkState(!closed, "SimpleLocalMemoryContext is already closed");
+
+        if (bytes == usedBytes) {
+            return NOT_BLOCKED;
+        }
+
+        // update the parent first as it may throw a runtime exception (e.g., ExceededMemoryLimitException)
+        ListenableFuture<?> future = parentMemoryContext.updateBytes(bytes - usedBytes);
+        usedBytes = bytes;
+        return future;
+    }
+
+    @Override
+    public synchronized boolean trySetBytes(long bytes)
+    {
+        checkState(!closed, "SimpleLocalMemoryContext is already closed");
+        long delta = bytes - usedBytes;
+        if (parentMemoryContext.tryUpdateBytes(delta)) {
+            usedBytes += delta;
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public synchronized void transferOwnership(LocalMemoryContext to)
+    {
+        checkArgument(to instanceof SimpleLocalMemoryContext, "to must be an instance of SimpleLocalMemoryContext");
```

Comment:
Looks like this method can be simplified (see below). As a side-effect, it would be coded to interface rather than a particular implementation, removing need for this check,